### PR TITLE
Add the possibility of using nested joins

### PR DESCRIPTION
**Pull request by richard-fizz**
Adds the posibility to use a 'join join' on the left side of the joined selector

Depends on
- https://github.com/phpcr/phpcr-utils/pull/160
- https://github.com/phpcr/phpcr-api-tests/pull/177

### CODE REVIEW

**Comment by lsmith77** [comment]
nice work ... but it would be important to validate this via a test in https://github.com/phpcr/phpcr-api-tests .. can you send a PR there with a test case?

**Comment by richard-fizz** [comment]
I've added the PR:
https://github.com/phpcr/phpcr-api-tests/pull/177
I'm fairly new to unit testing to be honest so it took me a while figuring it out :)

File: composer.json [hunk 1]
```diff
@@ -15,6 +15,12 @@
             "homepage": "https://github.com/jackalope/jackalope-jackrabbit/contributors"
         }
     ],
+    "repositories": [
+        {
+            "type": "vcs",
+            "url":  "https://github.com/richard-fizz/jackalope-doctrine-dbal.git"
```
Comment (dbu) [inline composer.json]: you need to do this with the phpcr-api-tests and the phpcr-utils instead of jackalope-doctrine-dbal. and change the version requirement on phpcr-api-tests to dev-master to get your changes in. same with phpcr-utils. (if you did the changes in a branch, it would be dev-<that-branch-name>)

**Comment by richard-fizz** [comment]
Of course... Apologies. The correct repositories are now added and set to pull from dev-master.

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php [hunk 2]
```diff
@@ -124,7 +124,7 @@ private function getSelectorAlias($selectorName)
     }
 
     /**
-     * @param QOM\QueryObjectModelInterface $qom
+     * @param QueryObjectModel $qom
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: @dantleech can you comment on this one please?

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php [hunk 3]
```diff
@@ -294,16 +294,25 @@ public function walkSelectorSource(QOM\SelectorInterface $source)
      *
      * @throws NotImplementedException
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: this comment can be updated :-) please also add an explanation what the $root parameter means.

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php [hunk 4]
```diff
@@ -345,7 +357,12 @@ public function walkJoinCondition(QOM\SelectorInterface $left, QOM\SelectorInter
         }
 
         if ($condition instanceof QOM\EquiJoinConditionInterface) {
-            return $this->walkEquiJoinCondition($left->getSelectorName(), $right->getSelectorName(), $condition);
+            if (!$left instanceof QOM\SelectorInterface) {
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: can we invert these for readability? (if $left instanceof then getSelectorName else getJoinCondition...)

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php [hunk 5]
```diff
@@ -387,8 +404,8 @@ public function walkDescendantNodeJoinCondition(QOM\DescendantNodeJoinConditionI
     public function walkEquiJoinCondition($leftSelectorName, $rightSelectorName, QOM\EquiJoinConditionInterface $condition)
     {
         return $this->walkOperand(new PropertyValue($leftSelectorName, $condition->getProperty1Name())) . " " .
-               $this->walkOperator(QOM\QueryObjectModelConstantsInterface::JCR_OPERATOR_EQUAL_TO) . " " .
-               $this->walkOperand(new PropertyValue($rightSelectorName, $condition->getProperty2Name()));
+        $this->walkOperator(QOM\QueryObjectModelConstantsInterface::JCR_OPERATOR_EQUAL_TO) . " " .
+        $this->walkOperand(new PropertyValue($rightSelectorName, $condition->getProperty2Name()));
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: afaik the indention was correct here, or should be 4 spaces but not 0.

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php [hunk 6]
```diff
@@ -448,7 +465,7 @@ public function walkSameNodeConstraint(QOM\SameNodeInterface $constraint)
      */
     public function walkFullTextSearchConstraint(QOM\FullTextSearchInterface $constraint)
     {
-        return $this->sqlXpathExtractValue($this->getTableAlias($constraint->getSelectorName()), $constraint->getPropertyName()).' LIKE '. $this->conn->quote('%'.$constraint->getFullTextSearchExpression().'%');
+        return $this->sqlXpathExtractValue($this->getTableAlias($constraint->getSelectorName()), $constraint->getPropertyName()) . ' LIKE ' . $this->conn->quote('%' . $constraint->getFullTextSearchExpression() . '%');
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: we have no spaces between the string concatenating dots.

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php (hunk 3 above)
Comment (richard-fizz) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: The method still throws a NotImplementedException when the right side of the join is not an instance of QOM\SelectorInterface. Will add an explanation for the $root parameter!

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php [hunk 7]
```diff
@@ -323,18 +332,21 @@ public function walkJoinSource(QOM\JoinInterface $source)
         $sql .= "AND " . $this->walkJoinCondition($source->getLeft(), $source->getRight(), $source->getJoinCondition()) . " ";
         $sql .= ") "; // close on-clause
 
-        $sql .= "WHERE $leftAlias.workspace_name = ? AND $leftAlias.type IN ('" . $source->getLeft()->getNodeTypeName() ."'";
-        $subTypes = $this->nodeTypeManager->getSubtypes($source->getLeft()->getNodeTypeName());
-        foreach ($subTypes as $subType) {
-            /* @var $subType \PHPCR\NodeType\NodeTypeInterface */
-            $sql .= ", '" . $subType->getName() . "'";
+
+        if ($root) {
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: should we add ADD conditions for the subtypes of the right side?

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php [hunk 8]
```diff
@@ -294,16 +294,25 @@ public function walkSelectorSource(QOM\SelectorInterface $source)
...
+        $this->source = $left = $source->getLeft(); // The $left variable is used for storing the leftmost selector
 
-        $this->source = $source->getLeft();
-        $leftAlias = $this->getTableAlias($source->getLeft()->getSelectorName());
-        $sql = "FROM phpcr_nodes $leftAlias ";
+        if (!$source->getRight() instanceof QOM\SelectorInterface) {
+            throw new NotImplementedException('The right side of the join should not consist of another join');
+        }
 
+        if (!$source->getLeft() instanceof QOM\SelectorInterface) {
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: please switch the cases to avoid the negation

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php (hunk 6 above)
Comment (richard-fizz) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: Apologies. My IDE reformatted the code. Will revert such cases if necessary.

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php (hunk 7 above)
Comment (richard-fizz) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: Haven't gone much into detail about this part to be honest, but obviously made sure the functionality stays the same. Would it be adequate to leave a TODO?

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php (hunk 3 above)
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: ah, then we should add that, saying `@throws NotImplementedException if the right side of a join is a nested join`

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php (hunk 7 above)
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: @dantleech is there an easy solution here? can we just add the type query for each part instead of only the leftmost?
if there is no obvious solution, then yes lets add a TODO, its the same behaviour we currently had, which seems unfinished but as long as everybody uses nt:unstructured everywhere is irrelevant anyways.

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php (hunk 2 above)
Comment (dantleech) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: Well, this is more accurate :) If were to change the signature of this method to the interface it should be a different PR.

Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: oh well. right :-)

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php [hunk 9]
```diff
@@ -289,21 +294,31 @@ public function walkSelectorSource(QOM\SelectorInterface $source)
 
     /**
      * @param QOM\JoinInterface $source
+     * @param boolean $root
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: can you please add an explanation of the semantics of the parameter? i know that this class is making a bad example with missing explanations, but we should work on improving the situation when we add new things.

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php [hunk 10]
```diff
@@ -323,18 +338,31 @@ public function walkJoinSource(QOM\JoinInterface $source)
...
+            $sql .= ')';
         }
-        $sql .= ')';
 
         return $sql;
     }
 
-    public function walkJoinCondition(QOM\SelectorInterface $left, QOM\SelectorInterface $right, QOM\JoinConditionInterface $condition)
+    /**
+     * @param $left
```
Comment (dbu) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: would this be the types QOM\SelectorInterface|QOM\JoinConditionInterface ?

**Comment by dbu** [comment]
looks like the functionality is good. i started the PR to validate the tests with jackalope-jackrabbit in https://github.com/jackalope/jackalope-jackrabbit/pull/131

File: src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php (hunk 10 above)
Comment (richard-fizz) [inline src/Jackalope/Transport/DoctrineDBAL/Query/QOMWalker.php]: No, QOM\SelectorInterface|QOM\JoinInterface, but I did indeed miss that part.

**Comment by richard-fizz** [comment]
It now suddenly fails on PHP 5.3 on pgsql? Did something change?

File: composer.json [hunk 11]
```diff
@@ -15,11 +15,21 @@
...
+            "type": "vcs",
+            "url":  "https://github.com/richard-fizz/phpcr-api-tests.git"
+        }
+    ],
     "require": {
         "php": ">=5.3.3",
         "doctrine/dbal": ">=2.4.5,<3.0.x-dev",
         "phpcr/phpcr": "~2.1.2",
-        "phpcr/phpcr-utils": "^1.2.8",
+        "phpcr/phpcr-utils": "dev-master",
```
Comment (dbu) [inline composer.json]: i merged and tagged this one. please remove the repositories entry for the utils and set this to `^1.2.9`

**Comment by dbu** [comment]
ups, this looks as if you had rebased upstream master on top of your branch, instead of your branch on upstream master. the command you should use for this situation is `git rebase upstream/master`. but don't worry, i can sort this out once everything is ready to merge. (if you want a longer background on merge vs rebase, read https://medium.com/@porteneuve/getting-solid-at-git-rebase-vs-merge-4fa1a48c53aa#.ybqgrs1ab )

**Comment by richard-fizz** [comment]
Hm I'm going to have to do some research into various git features I haven't worked with before... Sorry for constantly messing it up.
Is it possible for you to merge from here? Or should I go back and do it again?

**Comment by dbu** [comment]
i should be able to sort it out.

no worries, we appreciate the contributions! keep them coming, and we both learn new things about git ;-)

**Comment by dbu** [comment]
continued in #319
